karapace/protobuf/io.py

```diff
@@ -0,0 +1,248 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from io import BytesIO
+from karapace.protobuf.exception import IllegalArgumentException, ProtobufSchemaResolutionException, ProtobufTypeException
+from karapace.protobuf.message_element import MessageElement
+from karapace.protobuf.protobuf_to_dict import dict_to_protobuf, protobuf_to_dict
+from karapace.protobuf.schema import ProtobufSchema
+from karapace.protobuf.type_element import TypeElement
+
+import hashlib
+import importlib
+import importlib.util
+import logging
+import os
+
+ZERO_BYTE = b'\x00'
+
+logger = logging.getLogger(__name__)
+
+
+def calculate_class_name(name: str) -> str:
+    return "c_" + hashlib.md5(name.encode('utf-8')).hexdigest()
+
+
+class ProtobufDatumReader:
+    """Deserialize Protobuf-encoded data into a Python data structure."""
+
+    @staticmethod
+    def check_props(schema_one, schema_two, prop_list):
+        for prop in prop_list:
+            if getattr(schema_one, prop) != getattr(schema_two, prop):
+                return False
+        return True
+
+    @staticmethod
+    def match_schemas(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema) -> bool:
+        # TODO (serge): schema comparison by fields required
+
+        if str(writer_schema) == str(reader_schema):
+            return True
+        return False
+
+    def __init__(self, writer_schema=None, reader_schema=None):
+        """
+    As defined in the Protobuf specification, we call the schema encoded
+    in the data the "writer's schema", and the schema expected by the
+    reader the "reader's schema".
+    """
+        self._writer_schema = writer_schema
+        self._reader_schema = reader_schema
+
+    # read/write properties
+    def set_writer_schema(self, writer_schema):
+        self._writer_schema = writer_schema
+
+    writer_schema = property(lambda self: self._writer_schema, set_writer_schema)
+
+    def set_reader_schema(self, reader_schema):
+        self._reader_schema = reader_schema
+
+    reader_schema = property(lambda self: self._reader_schema, set_reader_schema)
+
+    @staticmethod
+    def read_varint(bio: BytesIO) -> int:
+        """Read a variable-length integer.
+
+        :returns: Integer
+        """
+        varint = 0
+        read_bytes = 0
+
+        while True:
+            char = bio.read(1)
+            if len(char) == 0:
```

Comment:
and this if condition can be shortened into `if len(char) == 0 and read_bytes = 0:`

Reply:
I plan to uncomment raise below.